### ISSUE TITLE
Use public readonly RobotEvents API key 

### ISSUE THREAD

**Issue by MayorMonty** [elapse-vrc/elapse-app#37]
> A migration to the RobotEvents API when we get a key

You will not want to have a single API key for all users of your application. This is because the ratelimit of the RobotEvents API v2 is calculated per API token, which means your users would be limited to 1080 requests per minute **total**. 

I've spoken a little with RECF staff about API v2, and for this sort of use case what you want to do is generate a temporary session token by making a request to `https://robotevents.com`. This session token is provided in the `Set-Cookie` header and resets about once a day. I've built-in automatic token generation and renewal to my [client library](https://github.com/MayorMonty/robotevents).

Once you have a session token, you can present that in your requests for access to the API.

Good Luck with your project!

**Comment by Nomnomburger** [comment]
Cool, thanks!

**Comment by Nomnomburger** [comment]
So basically I need to get a temporary key for each time a user accesses the app, and that will not affect any other users with the limit?

**Comment by MayorMonty** [comment]
Yes. All of the live updating pages on RobotEvents have been reprogrammed to use the API v2, and this is how they access that data.

**Comment by Nomnomburger** [comment]
Cool! I might take a look at your package and see how i can make it in Dart, if that's ok lol

**Comment by a-fyn** [comment]
it's definitely not done, lol

**Comment by Nomnomburger** [comment]
@MayorMonty so essentially the authentication for the data is run everytime the data is trying to be fetched, is that correct?

Also correct me if I'm wrong but the authentication and getting data works like:
1. Get token stored in the html `<header>` from `robotevents.com`
2. Take that token and put it in the cookies when making an HTTP request to the API
3. Get all of the data, and (for Dart & Flutter) parse the JSON data
4. Use the data

Pls let me know for the above, thanks for your help! :)

**Comment by MayorMonty** [comment]
No HTML parsing is required. The only thing you need is the value returned from the `set-cookie` header, specifically, `re_session`. 

```
export async function basic() {
  const response = await fetch("https://www.robotevents.com");

  if (!response.ok) {
    return Promise.reject(
      new Error("Could not authentication cookie")
    );
  }

  const cookie = parse(response.headers.raw()["set-cookie"]);

  // Invalidate the cookie after re_session expires
  const re_session = cookie.find((c) => c.name === "re_session");
  const expires = 1000 * (re_session?.maxAge ?? 0) + Date.now();

  return setCookie(
    cookie.map((c) => `${c.name}=${c.value}`).join("; "),
    expires
  );
}
```

Once you've extracted the cookie from your initial request, with every additional API request you make, add a `cookie` header with all of the cookies from `Set-Cookie`, and make sure to request a new set once they expire.

**Comment by Nomnomburger** [comment]
Thanks, I'll try.

btw how do you determine when the key has expired?

**Comment by MayorMonty** [comment]
When the cookies expire, you'll need to make another request to get a new session. This should be about every hour. 

Check out the MDN Reference for   [`Set-Cookie`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) for more info

**Comment by Nomnomburger** [comment]
@MayorMonty thanks.

So essentially, the code needs to:
1. get the re_session cookie from robotevents.com
2. apply the cookie into the desired api page with set-cookie
3. refresh the process when the current date/time is after the expires date/time on the cookie

i think its like that, right

**Comment by Nomnomburger** [comment]
Epic getting data now works
